samza-api/src/main/java/org/apache/samza/system/StreamSpec.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.system;
+
+import java.util.Properties;
+
+
+/**
+ * StreamSpec is a blueprint for creating, validating, or simply describing a stream in the runtime environment.
+ *
+ * It has specific attributes for common behaviors that samza uses.
+ *
+ * It also includes a set of properties which may be systemName-specific.
+ */
+public class StreamSpec {
+
+  /**
+   * Unique identifier for the stream in a Samza application.
+   * This identifier is used as a key for stream properties in
+   * config and to distinguish between streams in a graph.
+   */
+  private final String id;
+
+  /**
+   * The System name on which this stream will exist. Corresponds to a named implementation of the
+   * Samza System abstraction.
+   */
+  private final String systemName;
+
+  /**
+   * The physical identifier for the stream. This is the identifier that will be used in remote
+   * systems to identify the stream. In Kafka this would be the topic name whereas in HDFS it
+   * might be a file URN.
+   */
+  private String physicalName;
+
+  /**
+   * The number of partitions for the stream.
+   */
+  private int partitionCount = 1;
```

Comment:
Given how close we are to an immutable data structure, I would be OK with that tradeoff - especially since planning is not going to be an ongoing cost we pay while actually running the job. I'm pretty heavily biased in that direction though.